text/0000-type-declarations.md

```diff
@@ -0,0 +1,219 @@
+
+```php
+protected function foo( array $data, WP_REST_Request $request ) {
+	// ...
+}
+```
+
+### Required Form
+
+_**Note:** Use `int`, not `integer`. Use `bool`, not `boolean`._
+
+The following are valid Type Declarations:
+
+- Class or interface name; e.g., `WP_REST_Request`
+- `self`, which references own class or interface name
+- `parent`, which references parent class or interface name
+- `array`
+- `callable`
+- `string`
+- `int`
+- `float`
+- `bool`
+- `iterable`
+- `object`
+- `void`, as a Return Type Declaration
+
+## WordPress Core Compatibility
+
+At this time, the additional Type Declarations: `callable`, `string`, `int`, `float`, `bool`, `iterable`, and `object` must be avoided in WordPress Core. The same is true for [Return Type Declarations](http://php.net/manual/en/migration70.new-features.php#migration70.new-features.return-type-declarations), [Strict Typing](http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration.strict), and [Nullable Types](http://php.net/manual/en/migration71.new-features.php#migration71.new-features.nullable-types). Please see [WordPress requirements](https://wordpress.org/about/requirements/) (PHP 5.2.4+) and [this table](http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration) for further details.
+
+The only Type Declarations supported in WordPress Core at this time, are:
+
+- Class or interface name; e.g., `WP_REST_Request`
+- `self`, which references own class or interface name
+- `parent`, which references parent class or interface name
+- `array`, to require an array
+
+_These can only be used in function parameters, not as Return Type Declarations._
+
+## When to Use Type Declarations
+
+When you're writing a function or a class method that expects to receive an array or a specific object type, and there is no reason to accept anything other than that specific type.
+
+```php
+protected function foo( Bar $bar, array $items ) {
+	// $bar   must be an instance of the Bar class.
+	// $items must be passed as an array.
+}
+```
+
+## When Not to Use Type Declarations
+
+If the function you're writing is part of a public API and enforcing a specific data type would make the function less flexible in the eyes of a caller. For example, the [`get_post()`](https://developer.wordpress.org/reference/functions/get_post/) function in WordPress core accepts `int|WP_Post|null` as the first parameter. This maximizes flexibility for callers, making the function more convenient in a variety of circumstances.
+
+Likewise, if you're writing a public function for an API and it needs a `WP_Post` instance, it's better not to enforce `WP_Post` with a type declaration. Instead, use the `get_post()` function to resolve the `$post` reference, making it possible for a caller to pass `int|WP_Post|null` to your function as well.
+
+```php
+function foo( $post ) {
+	$post = get_post( $post );
+	...
+}
+```
+
```

Comment:
Given the current PHP requirements, I think that type declarations should only be used right now with types that are:
- interfaces or classes
- considered to be part of the public interface
 - extensible

All other type declarations should be discouraged, as they cause more problems than they solve:
- Using `array` immediately renders almost any future improvements using smart objects impossible. It makes it impossible for objects to stand in as collections of elements. Often times, when the doc-block states that a function needs an `array`, it actually means it needs a `Traversable`, because it wants to loop over it, but cannot do that as PHP 5.2 is missing that concept, or it needs an `ArrayAccess` because it wants to use indexed access. Using `array` should be avoided at all costs, unless there is a specific reason why it needs to be the actual internal type of `array`.
- Using classes that are not extensible means that future changes need to be done by _changing_  an element of the public interface, instead of _ extending_  it. This is again an unneeded hindrance to flexibility. This is why _segregated interfaces_ (i.e. `Renderable` or `Taggable` instead of `WP_Post`) should always be preferred, however WordPress is not really big on using interfaces.

Reply:
Due to the above, I'd also recommend changing the examples in this document to not so prominently show the `array` type declaration.